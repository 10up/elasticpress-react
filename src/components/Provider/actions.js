export const SET_RESULTS = 'SET_RESULTS';
export const SET_LOADING = 'SET_LOADING';
export const SET_SEARCH_TERMS = 'SET_SEARCH_TERMS';
export const REFINE = 'REFINE';
