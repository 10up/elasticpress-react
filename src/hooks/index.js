export { default as useElasticPress } from './useElasticPress';
export { default as useRoveFocus } from './useRoveFocus';
export { default as useSearch } from './useSearch';
