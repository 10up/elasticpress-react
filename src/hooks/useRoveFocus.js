import { useState, useCallback, useEffect } from 'react';

export default function useRoveFocus(size) {
	const [currentFocus, setCurrentFocus] = useState(-1);

	const handleKeyDown = useCallback(
		(e) => {
			if (e.keyCode === 40) {
				// Down arrow
				e.preventDefault();
				setCurrentFocus(currentFocus === size - 1 ? -1 : currentFocus + 1);
			} else if (e.keyCode === 38) {
				// Up arrow
				e.preventDefault();
				setCurrentFocus(currentFocus === -1 ? size - 1 : currentFocus - 1);
			}
		},
		[size, currentFocus, setCurrentFocus],
	);

	useEffect(() => {
		document.addEventListener('keydown', handleKeyDown, false);
		return () => {
			document.removeEventListener('keydown', handleKeyDown, false);
		};
	}, [handleKeyDown]);

	return [currentFocus, setCurrentFocus];
}
