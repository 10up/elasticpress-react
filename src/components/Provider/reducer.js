import { SET_RESULTS, SET_LOADING, SET_SEARCH_TERMS } from './actions';
import { query as esQuery } from '../SearchField';

export const initialState = {
	results: null,
	searchTerms: null,
	perPage: 10,
	offset: 0,
	totalResults: null,
	loading: false,
	query: esQuery,
	hitMap: (hit) => {
		return hit._source;
	},
	node: null,
	indexName: null,
};

const reducer = (state, action) => {
	switch (action.type) {
		case SET_RESULTS: {
			const { append, ...payload } = action.payload;
			return {
				...state,
				...payload,
				results: append ? state.results.concat(payload.results) : payload.results,
			};
		}
		case SET_LOADING:
			return {
				...state,
				loading: action.payload,
			};
		case SET_SEARCH_TERMS:
			return {
				...state,
				searchTerms: action.payload,
			};
		default:
			return state;
	}
};

export default reducer;
