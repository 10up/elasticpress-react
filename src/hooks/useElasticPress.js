import { useContext } from 'react';
import { ElasticPressContext } from '../components/Provider/ElasticPressProvider';

/**
 * A simple hook to return the ElasticPressProvider state
 *
 * @returns {object}
 */
const useElasticPress = () => {
	return useContext(ElasticPressContext);
};

export default useElasticPress;
