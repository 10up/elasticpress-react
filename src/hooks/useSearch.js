import { useCallback } from 'react';

import useElasticPress from './useElasticPress';
import { SET_SEARCH_TERMS, SET_LOADING, SET_RESULTS } from '../components/Provider';
import { replacePlaceholderInObjectValues } from '../utils';
import { post } from '../api';

const useSearch = () => {
	const { dispatch, state, getEndpoint } = useElasticPress();

	const refine = useCallback(
		(value, minSearchCharacters) => {
			dispatch({
				type: SET_SEARCH_TERMS,
				payload: value,
			});

			if (value < minSearchCharacters) {
				return;
			}

			dispatch({
				type: SET_LOADING,
				payload: true,
			});

			let newQuery = replacePlaceholderInObjectValues(state.query, '%SEARCH_TERMS%', value);

			newQuery = replacePlaceholderInObjectValues(newQuery, '%PER_PAGE%', state.perPage);

			post(newQuery, getEndpoint('search')).then((response) => {
				let newResults = [];
				let totalResults = 0;

				if (response.hits && response.hits.hits) {
					if (response.hits.total && response.hits.total.value) {
						totalResults = parseInt(response.hits.total.value, 10);
					}
					newResults = response.hits.hits.map(state.hitMap);
				}

				dispatch({
					type: SET_RESULTS,
					payload: {
						results: newResults,
						totalResults,
						offset: state.perPage,
						value,
					},
				});

				dispatch({
					type: SET_LOADING,
					payload: false,
				});
			});
		},
		[dispatch, state.query, state.hitMap, state.perPage, getEndpoint],
	);

	return { refine, searchTerms: state.searchTerms };
};

export default useSearch;
