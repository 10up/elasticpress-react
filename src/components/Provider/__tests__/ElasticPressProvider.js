import React from 'react';
import { render } from '@testing-library/react';
import { renderHook, act } from '@testing-library/react-hooks';
import { ElasticPressProvider } from '..';
import { initialState } from '../reducer';
import { SET_SEARCH_TERMS } from '../actions';
import { useElasticPress } from '../../../hooks';

const AppTest = () => {
	return <p>I am a very simple test app</p>;
};

const node = 'http://elasticpress.test/__elasticsearch';
const indexName = 'test';

// eslint-disable-next-line react/prop-types
const ElasticPressProviderWrapper = ({ children }) => {
	return (
		<ElasticPressProvider node={node} indexName={indexName}>
			{children}
		</ElasticPressProvider>
	);
};

describe('ElasticPressProvider', () => {
	it('works', () => {
		const { container } = render(<AppTest />, { wrapper: ElasticPressProviderWrapper });

		expect(container).toMatchSnapshot();
	});

	it('initial state is equal to default state', () => {
		const { result } = renderHook(() => useElasticPress(), {
			wrapper: ElasticPressProviderWrapper,
		});

		expect(result.current.state).toMatchObject({ ...initialState, node, indexName });
	});

	it('only triggers a re-render if state changes', () => {
		const { result } = renderHook(() => useElasticPress(), {
			wrapper: ElasticPressProviderWrapper,
		});

		const originalState = result.current;

		// dispatching an action that does not change state should maintain object reference
		act(() => {
			result.current.dispatch({ type: 'dummy-action' });
		});

		expect(originalState === result.current).toBeTruthy();

		act(() => {
			result.current.dispatch({ type: 'dummy-action2' });
		});

		expect(originalState === result.current).toBeTruthy();

		// let's make sure dispatching an action that does change state triggers a re-render
		act(() => {
			result.current.dispatch({ type: SET_SEARCH_TERMS, payload: null });
		});

		expect(originalState === result.current).not.toBeTruthy();
		expect(originalState).toMatchObject(result.current);
	});
});
