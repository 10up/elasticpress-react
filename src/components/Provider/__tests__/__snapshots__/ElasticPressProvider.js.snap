// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticPressProvider works 1`] = `
<div>
  <p>
    I am a very simple test app
  </p>
</div>
`;
